File: neon_tetris.py
# neon_tetris.py
import curses
import random
import time
import os

# Константы
WIDTH = 10
HEIGHT = 20
SHAPES = [
    [[1, 1, 1, 1]],
    [[1, 1], [1, 1]],
    [[0, 1, 0], [1, 1, 1]],
    [[1, 0, 0], [1, 1, 1]],
    [[0, 0, 1], [1, 1, 1]],
    [[0, 1, 1], [1, 1, 0]],
    [[1, 1, 0], [0, 1, 1]]
]
COLORS = [1, 2, 3, 4, 5, 6, 7]  # цветовые пары curses
COLOR_NAMES = ['cyan', 'yellow', 'magenta', 'green', 'red', 'blue', 'white']

class Tetris:
    def __init__(self, stdscr):
        self.stdscr = stdscr
        curses.curs_set(0)
        curses.start_color()
        curses.init_pair(1, curses.COLOR_CYAN, curses.COLOR_BLACK)
        curses.init_pair(2, curses.COLOR_YELLOW, curses.COLOR_BLACK)
        curses.init_pair(3, curses.COLOR_MAGENTA, curses.COLOR_BLACK)
        curses.init_pair(4, curses.COLOR_GREEN, curses.COLOR_BLACK)
        curses.init_pair(5, curses.COLOR_RED, curses.COLOR_BLACK)
        curses.init_pair(6, curses.COLOR_BLUE, curses.COLOR_BLACK)
        curses.init_pair(7, curses.COLOR_WHITE, curses.COLOR_BLACK)
        self.stdscr.keypad(True)
        self.stdscr.nodelay(True)
        self.board = [[0] * WIDTH for _ in range(HEIGHT)]
        self.score = 0
        self.lines = 0
        self.level = 1
        self.fall_time = 500  # мс
        self.last_fall = time.time() * 1000
        self.game_over = False
        self.paused = False
        self.current_piece = None
        self.next_piece = None
        self.spawn_piece()

    def spawn_piece(self):
        self.current_piece = self.next_piece if self.next_piece else self.new_piece()
        self.next_piece = self.new_piece()
        self.piece_x = WIDTH // 2 - len(self.current_piece[0]) // 2
        self.piece_y = 0
        if self.collision(self.current_piece, self.piece_x, self.piece_y):
            self.game_over = True

    def new_piece(self):
        shape = random.choice(SHAPES)
        color = random.choice(COLORS)
        return {'shape': shape, 'color': color}

    def collision(self, piece, x, y):
        for row in range(len(piece['shape'])):
            for col in range(len(piece['shape'][row])):
                if piece['shape'][row][col]:
                    board_x = x + col
                    board_y = y + row
                    if board_x < 0 or board_x >= WIDTH or board_y >= HEIGHT:
                        return True
                    if board_y >= 0 and self.board[board_y][board_x]:
                        return True
        return False

    def lock_piece(self):
        piece = self.current_piece
        for row in range(len(piece['shape'])):
            for col in range(len(piece['shape'][row])):
                if piece['shape'][row][col]:
                    y = self.piece_y + row
                    x = self.piece_x + col
                    if y >= 0:
                        self.board[y][x] = piece['color']
        self.clear_lines()
        self.spawn_piece()

    def clear_lines(self):
        lines_cleared = 0
        for y in range(HEIGHT-1, -1, -1):
            if all(self.board[y]):
                del self.board[y]
                self.board.insert(0, [0] * WIDTH)
                lines_cleared += 1
        if lines_cleared:
            self.lines += lines_cleared
            self.score += lines_cleared * 100
            self.level = self.lines // 10 + 1
            self.fall_time = max(100, 500 - (self.level - 1) * 30)

    def move(self, dx, dy):
        new_x = self.piece_x + dx
        new_y = self.piece_y + dy
        if not self.collision(self.current_piece, new_x, new_y):
            self.piece_x = new_x
            self.piece_y = new_y
            return True
        if dy == 1:  # падение вниз не удалось – фиксируем
            self.lock_piece()
        return False

    def rotate(self):
        shape = self.current_piece['shape']
        rotated = list(zip(*shape[::-1]))
        new_piece = {'shape': [list(row) for row in rotated], 'color': self.current_piece['color']}
        if not self.collision(new_piece, self.piece_x, self.piece_y):
            self.current_piece = new_piece
        else:
            # попытка "стены"
            for dx in [-1, 1]:
                if not self.collision(new_piece, self.piece_x + dx, self.piece_y):
                    self.piece_x += dx
                    self.current_piece = new_piece
                    break

    def drop(self):
        while not self.collision(self.current_piece, self.piece_x, self.piece_y + 1):
            self.piece_y += 1
        self.lock_piece()

    def draw(self):
        self.stdscr.clear()
        # Рамка
        h, w = self.stdscr.getmaxyx()
        board_win = curses.newwin(HEIGHT+2, WIDTH+2, 2, 2)
        board_win.border(0)
        for y in range(HEIGHT):
            for x in range(WIDTH):
                if self.board[y][x]:
                    board_win.addch(y+1, x+1, '█', curses.color_pair(self.board[y][x]))
        # Текущая фигура
        piece = self.current_piece
        for row in range(len(piece['shape'])):
            for col in range(len(piece['shape'][row])):
                if piece['shape'][row][col]:
                    by = self.piece_y + row + 1
                    bx = self.piece_x + col + 1
                    if by >= 1 and by <= HEIGHT:
                        board_win.addch(by, bx, '█', curses.color_pair(piece['color']))
        board_win.refresh()
        # Информация
        info = f"Score: {self.score}  Level: {self.level}  Lines: {self.lines}"
        self.stdscr.addstr(1, 2, info)
        self.stdscr.addstr(1, 30, "Next:")
        # Показываем следующую фигуру
        next_shape = self.next_piece['shape']
        for r in range(len(next_shape)):
            for c in range(len(next_shape[r])):
                if next_shape[r][c]:
                    self.stdscr.addch(2+r, 36+c, '█', curses.color_pair(self.next_piece['color']))
        if self.paused:
            self.stdscr.addstr(HEIGHT//2, WIDTH//2 - 2, "PAUSED", curses.A_REVERSE)
        if self.game_over:
            self.stdscr.addstr(HEIGHT//2, WIDTH//2 - 4, "GAME OVER", curses.A_REVERSE)
        self.stdscr.refresh()

    def run(self):
        while not self.game_over:
            now = time.time() * 1000
            # Ввод
            key = self.stdscr.getch()
            if key == ord('q'):
                break
            if key == ord('p'):
                self.paused = not self.paused
            if not self.paused:
                if key == curses.KEY_LEFT:
                    self.move(-1, 0)
                elif key == curses.KEY_RIGHT:
                    self.move(1, 0)
                elif key == curses.KEY_DOWN:
                    self.move(0, 1)
                elif key == curses.KEY_UP:
                    self.rotate()
                elif key == ord(' '):
                    self.drop()
                # Падение по таймеру
                if now - self.last_fall > self.fall_time:
                    self.move(0, 1)
                    self.last_fall = now
            self.draw()
            time.sleep(0.02)
        self.stdscr.getch()

def main(stdscr):
    game = Tetris(stdscr)
    game.run()

if __name__ == '__main__':
    curses.wrapper(main)

File: neon_tetris.cpp
// neon_tetris.cpp
#include <iostream>
#include <curses.h>
#include <unistd.h>
#include <cstdlib>
#include <ctime>
#include <vector>
#include <algorithm>

using namespace std;

const int WIDTH = 10;
const int HEIGHT = 20;

// Фигуры (матрицы)
vector<vector<vector<int>>> SHAPES = {
    {{1,1,1,1}},
    {{1,1},{1,1}},
    {{0,1,0},{1,1,1}},
    {{1,0,0},{1,1,1}},
    {{0,0,1},{1,1,1}},
    {{0,1,1},{1,1,0}},
    {{1,1,0},{0,1,1}}
};

// Цветовые пары
int COLORS[] = {1,2,3,4,5,6,7};

class Tetris {
public:
    Tetris() {
        initscr();
        start_color();
        init_pair(1, COLOR_CYAN, COLOR_BLACK);
        init_pair(2, COLOR_YELLOW, COLOR_BLACK);
        init_pair(3, COLOR_MAGENTA, COLOR_BLACK);
        init_pair(4, COLOR_GREEN, COLOR_BLACK);
        init_pair(5, COLOR_RED, COLOR_BLACK);
        init_pair(6, COLOR_BLUE, COLOR_BLACK);
        init_pair(7, COLOR_WHITE, COLOR_BLACK);
        cbreak();
        noecho();
        keypad(stdscr, TRUE);
        curs_set(0);
        nodelay(stdscr, TRUE);
        timeout(20);
        srand(time(nullptr));
        board.assign(HEIGHT, vector<int>(WIDTH, 0));
        score = 0; lines = 0; level = 1;
        fallTime = 500;
        lastFall = clock();
        gameOver = false; paused = false;
        nextPiece = randomPiece();
        spawnPiece();
    }

    ~Tetris() {
        endwin();
    }

    void run() {
        while (!gameOver) {
            if (!paused) {
                clock_t now = clock();
                if ((now - lastFall) * 1000 / CLOCKS_PER_SEC > fallTime) {
                    movePiece(0, 1);
                    lastFall = now;
                }
            }
            draw();
            handleInput();
            usleep(20000);
        }
        mvprintw(HEIGHT/2, WIDTH/2 - 4, "GAME OVER");
        refresh();
        getch();
    }

private:
    vector<vector<int>> board;
    vector<vector<int>> currentShape;
    vector<vector<int>> nextShape;
    int currentColor, nextColor;
    int pieceX, pieceY;
    int score, lines, level;
    int fallTime;
    clock_t lastFall;
    bool gameOver, paused;

    vector<vector<int>> randomPiece() {
        int idx = rand() % SHAPES.size();
        return SHAPES[idx];
    }

    void spawnPiece() {
        currentShape = nextShape;
        currentColor = nextColor;
        nextShape = randomPiece();
        nextColor = rand() % 7 + 1;
        pieceX = WIDTH / 2 - currentShape[0].size() / 2;
        pieceY = 0;
        if (collision(currentShape, pieceX, pieceY)) gameOver = true;
    }

    bool collision(const vector<vector<int>>& shape, int x, int y) {
        for (int row = 0; row < shape.size(); row++) {
            for (int col = 0; col < shape[row].size(); col++) {
                if (shape[row][col]) {
                    int bx = x + col;
                    int by = y + row;
                    if (bx < 0 || bx >= WIDTH || by >= HEIGHT) return true;
                    if (by >= 0 && board[by][bx]) return true;
                }
            }
        }
        return false;
    }

    void lockPiece() {
        for (int row = 0; row < currentShape.size(); row++) {
            for (int col = 0; col < currentShape[row].size(); col++) {
                if (currentShape[row][col]) {
                    int by = pieceY + row;
                    int bx = pieceX + col;
                    if (by >= 0) board[by][bx] = currentColor;
                }
            }
        }
        clearLines();
        spawnPiece();
    }

    void clearLines() {
        int cleared = 0;
        for (int y = HEIGHT - 1; y >= 0; y--) {
            bool full = true;
            for (int x = 0; x < WIDTH; x++) if (!board[y][x]) { full = false; break; }
            if (full) {
                for (int ny = y; ny > 0; ny--) board[ny] = board[ny-1];
                board[0].assign(WIDTH, 0);
                cleared++;
                y++;
            }
        }
        if (cleared) {
            lines += cleared;
            score += cleared * 100;
            level = lines / 10 + 1;
            fallTime = max(100, 500 - (level - 1) * 30);
        }
    }

    void movePiece(int dx, int dy) {
        int nx = pieceX + dx;
        int ny = pieceY + dy;
        if (!collision(currentShape, nx, ny)) {
            pieceX = nx;
            pieceY = ny;
            return;
        }
        if (dy == 1) lockPiece();
    }

    void rotate() {
        int rows = currentShape.size();
        int cols = currentShape[0].size();
        vector<vector<int>> rotated(cols, vector<int>(rows));
        for (int i = 0; i < rows; i++)
            for (int j = 0; j < cols; j++)
                rotated[j][rows - 1 - i] = currentShape[i][j];
        if (!collision(rotated, pieceX, pieceY)) {
            currentShape = rotated;
        } else {
            for (int dx : {-1, 1}) {
                if (!collision(rotated, pieceX + dx, pieceY)) {
                    pieceX += dx;
                    currentShape = rotated;
                    break;
                }
            }
        }
    }

    void drop() {
        while (!collision(currentShape, pieceX, pieceY + 1)) pieceY++;
        lockPiece();
    }

    void draw() {
        clear();
        // рамка
        mvprintw(0, 0, "+");
        for (int i = 0; i < WIDTH; i++) addch('-');
        addch('+');
        for (int y = 0; y < HEIGHT; y++) {
            mvaddch(y+1, 0, '|');
            for (int x = 0; x < WIDTH; x++) {
                int color = 0;
                char ch = ' ';
                if (board[y][x]) {
                    color = board[y][x];
                    ch = '█';
                }
                // current
                for (int row = 0; row < currentShape.size(); row++) {
                    for (int col = 0; col < currentShape[row].size(); col++) {
                        if (currentShape[row][col]) {
                            int by = pieceY + row;
                            int bx = pieceX + col;
                            if (by == y && bx == x) {
                                color = currentColor;
                                ch = '█';
                            }
                        }
                    }
                }
                if (ch != ' ') {
                    attron(COLOR_PAIR(color));
                    mvaddch(y+1, x+1, ch);
                    attroff(COLOR_PAIR(color));
                } else mvaddch(y+1, x+1, ' ');
            }
            mvaddch(y+1, WIDTH+1, '|');
        }
        mvprintw(HEIGHT+1, 0, "+");
        for (int i = 0; i < WIDTH; i++) addch('-');
        addch('+');
        mvprintw(HEIGHT+2, 0, "Score: %d  Level: %d  Lines: %d", score, level, lines);
        if (paused) mvprintw(HEIGHT/2, WIDTH/2 - 3, "PAUSED");
        if (gameOver) mvprintw(HEIGHT/2, WIDTH/2 - 4, "GAME OVER");
        refresh();
    }

    void handleInput() {
        int ch = getch();
        if (ch == 'q') { gameOver = true; return; }
        if (ch == 'p') { paused = !paused; return; }
        if (paused) return;
        switch (ch) {
            case KEY_LEFT: movePiece(-1, 0); break;
            case KEY_RIGHT: movePiece(1, 0); break;
            case KEY_DOWN: movePiece(0, 1); break;
            case KEY_UP: rotate(); break;
            case ' ': drop(); break;
        }
    }
};

int main() {
    Tetris game;
    game.run();
    return 0;
}
